### PR TITLE
multifvm

### PR DESCRIPTION
**Pull request by vyzo**
This adds support for multiple fvm versions behind the same facade to the client, with dispatch based on the network version.
It allows us to host multiple network versions in the same client, including in development things, without blocking ourselves on backwards compatibility.

### CODE REVIEW

File: rust/src/fvm/engine.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+use anyhow::anyhow;
+use cid::Cid;
+
+use fvm2::call_manager::DefaultCallManager as DefaultCallManager2;
```
Comment (Stebalien) [inline rust/src/fvm/engine.rs:8]: Can we split this into an `v2` and an `v3`? That way we wouldn't need to rename all of our imports.
I.e.:

```rust
mod engine {
    mod v2 { ... }
    mod v3 { ... }
    pub use v2::new_executor as new_executor2;
    pub use v3::new_executor as new_executor3;
}
```

File: rust/src/fvm/engine.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        raw_length: usize,
+    ) -> anyhow::Result<ApplyRet> {
+        use fvm2::executor::Executor;
+        let res = self.0.execute_message(
+            Message2 {
+                version: msg.version,
+                from: Address2::from_bytes(&msg.from.to_bytes()).unwrap(),
+                to: Address2::from_bytes(&msg.to.to_bytes()).unwrap(),
+                sequence: msg.sequence,
+                value: unsafe { std::mem::transmute::<TokenAmount3, TokenAmount2>(msg.value) },
```
Comment (Stebalien) [inline rust/src/fvm/engine.rs:107]: can we avoid this? It won't really cost anything.

Comment (Stebalien) [reply rust/src/fvm/engine.rs:107]: Yeah, let's get rid of the transmutes. We won't notice if something changes.

**Review by Stebalien — COMMENTED** [review]
Approach LGTM. One comment about modules, one about transmutes.

**Comment by Stebalien** [comment]
So, FUN TIMES! It looks like wasmtime is exporting some symbols, so we can't include it twice (in it's current incarnation, that is).

So... I think we need to:

1. Upgrade fvm v2 to wasmtime 1.0.
2. Test it _extensively_ on mainnet to make sure it's completely compatible.

https://github.com/filecoin-project/ref-fvm/issues/962

**Comment by Stebalien** [comment]
Depends on https://github.com/filecoin-project/ref-fvm/pull/963.

File: rust/src/fvm/engine.rs (hunk 2 above)
Comment (vyzo) [reply rust/src/fvm/engine.rs]: removed all the transmutes; this was painful, but much better than have a random (maybe silent) failure.

Obligatory comment about rust: it is insane that the layout of structs and stuff is not deterministic. This is the biggest advantage of C, you can rely on things being the same and laid out the same.

Comment (vmx) [reply rust/src/fvm/engine.rs]: Obligatory reply to rust comment :) If you want C semantics, you can annotate the structs with [`#[repr(C)]` ](https://doc.rust-lang.org/reference/type-layout.html#the-c-representation)`.

File: rust/src/fvm/engine.rs (hunk 1 above)
Comment (vyzo) [reply rust/src/fvm/engine.rs]: done; we don't even need to use the executors, it is handled internally.

**Comment by vyzo** [comment]
ok, this is done; rebasing on master and marking as ready for review.

File: rust/src/fvm/engine.rs (hunk 2 above)
Comment (vyzo) [reply rust/src/fvm/engine.rs]: I dont necessarily want C repr, i jist want the compiler to be sane and predictable damnit!

Side note: i have somewhat warmed to that language of yours, but i still get fits from all those little things.

**Review by Stebalien — APPROVED** [review]
I am so, SO sorry. This looks like hell.

File: rust/src/fvm/engine.rs [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+}
+
+impl Default for MultiEngineContainer {
+    fn default() -> MultiEngineContainer {
+        MultiEngineContainer::new()
+    }
+}
+
+// fvm v3 implementation
+mod v3 {
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply rust/src/fvm/engine.rs:81]: Eh, we can do that in a followup.

**Comment by Stebalien** [comment]
Ok, so, this doesn't work on mainnet.

**Comment by vyzo** [comment]
what happened?

**Comment by Stebalien** [comment]
> what happened?

General "state mismatch" errors. Let me get you access to the box.

**Comment by Stebalien** [comment]
Note: I've now tried FVM 2.1 and that's not the issue. Now I'm trying the few patches I just pushed, just in case.

**Comment by Stebalien** [comment]
I've now checked with the patches I just pushed, and they don't help.

**Comment by Stebalien** [comment]
I've confirmed that:

- Messages aren't just flat-out failing.
- The network config that we end up passing into the machine looks sane.

Looking at the diffs... I'm seeing _mostly_ state mismatches, but also some value mismatches. Strangely, I'm not just seeing random messages failing, so more debugging required...

**Comment by Stebalien** [comment]
I found that we weren't invoking the threaded executor, so I fixed that. But that hasn't changed the situation.

**Comment by Stebalien** [comment]
We will never mention this again.

**Comment by vyzo** [comment]
ouch, that was not the right randomness.... sorry bout that!

**Pull request merged**